exp/clusterresourceset/controllers/clusterresourceset_helpers.go

```diff
@@ -0,0 +1,174 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"io"
+
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	utilyaml "k8s.io/apimachinery/pkg/util/yaml"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	clusterresourcesetv1 "sigs.k8s.io/cluster-api/exp/clusterresourceset/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/yaml"
+)
+
+// ToUnstructured takes a YAML and converts it to a list of Unstructured objects
+func ToUnstructured(rawyaml []byte) ([]unstructured.Unstructured, error) {
+	var ret []unstructured.Unstructured //nolint
+
+	reader := utilyaml.NewYAMLReader(bufio.NewReader(bytes.NewReader(rawyaml)))
+	count := 1
+	for {
+		// Read one YAML document at a time, until io.EOF is returned
+		b, err := reader.Read()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return nil, errors.Wrapf(err, "failed to read yaml")
+		}
+		if len(b) == 0 {
+			break
+		}
+
+		var m map[string]interface{}
+		if err := yaml.Unmarshal(b, &m); err != nil {
+			return nil, errors.Wrapf(err, "failed to unmarshal the %s yaml document: %q", util.Ordinalize(count), string(b))
+		}
+
+		var u unstructured.Unstructured
+		u.SetUnstructuredContent(m)
+
+		// Ignore empty objects.
+		// Empty objects are generated if there are weird things in manifest files like e.g. two --- in a row without a yaml doc in the middle
+		if u.Object == nil {
+			continue
+		}
+
+		ret = append(ret, u)
+		count++
+	}
+
+	return ret, nil
+}
+
+func Apply(ctx context.Context, c client.Client, data []byte) error {
+	// Transform the yaml in a list of objects, so following transformation can work on typed objects (instead of working on a string/slice of bytes)
+	reader := bufio.NewReader(bytes.NewReader(data))
+	// TODO: Check how many bytes enough here
+	_, _, isJSON := utilyaml.GuessJSONStream(reader, 2048)
+	if isJSON {
+		var err error
+		data, err = yaml.JSONToYAML(data)
+		if err != nil {
+			return errors.Wrapf(err, "failed converting JSON data to YAML")
+		}
+	}
+	objs, err := ToUnstructured(data)
+	if err != nil {
+		return errors.Wrapf(err, "failed converting data to unstructured objects")
+	}
+
+	// TODO: if one of them fails, return with error or continue creating other objects in the resource?
+	for i := range objs {
+		err = apply(ctx, c, &objs[i])
```

Comment:
AFAIK, we should respect order in creating objects (e.g create crd & namespace as first etc etc)
this is how this is implemented in clusterctl https://github.com/kubernetes-sigs/cluster-api/blob/b8026b35a857fb15407290507cb3ae202fa581b6/cmd/clusterctl/client/cluster/components.go#L221-L275;
probably, we should move this into a util

Reply:
Yes, we can do this in a follow up PR.

Reply:
Moved this to util package.